compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -2298,17 +2298,33 @@ fn confirm_builtin_candidate<'cx, 'tcx>(
                 ty,
                 &mut obligations,
             )
-        });
-        if check_is_sized {
-            let sized_predicate = ty::TraitRef::from_lang_item(
-                tcx,
-                LangItem::Sized,
-                obligation.cause.span(),
-                [self_ty],
-            );
-            obligations.push(obligation.with(tcx, sized_predicate));
-        }
-        (metadata_ty.into(), obligations)
+        };
+        let metadata = match self_ty.ptr_metadata_ty_or_tail(tcx, normalize) {
+            Ok(metadata) => metadata,
+            Err(tail) => {
+                let sized_predicate = ty::TraitRef::from_lang_item(
+                    tcx,
+                    LangItem::Sized,
+                    obligation.cause.span(),
+                    [self_ty],
+                );
+                let sized_obligation = obligation.with(tcx, sized_predicate);
+                if self_ty == tail
+                    || selcx.infcx.predicate_must_hold_considering_regions(&sized_obligation)
```

Comment:
using `must_hold_considering_regions` outside of pure comp time optimizations should be avoided.

It results in weird changes between `Foo<'a>` and `Foo`. We should document this somewhere